### DISCUSSION TITLE
Viam MicroRDK Configuration

### DISCUSSION THREAD

**Discussion by danic85** [makerforgetech/modular-biped#78] [Viam]
Following these instructions:

https://docs.viam.com/get-started/installation/prepare/microcontrollers/

I have downloaded both of the files in `Setup` tab to a directory within `Downloads` on Ubuntu, then in the CLI, navigated to that directory and run the suggested install command.

`chmod 755 ./micro-rdk-installer-amd64-linux && ./micro-rdk-installer-amd64-linux write-flash --app-config viam-esp32test-main.json --monitor`

The first time it ran it asked if I wanted to use the  port `ttyS4` because the ESP32 wasn't connected yet. So I said 'no' and it exited.

The second time it didn't ask for a port, but I got the following error output:

```
$ viam-micro-rdk chmod 755 ./micro-rdk-installer-amd64-linux && ./micro-rdk-installer-amd64-linux write-flash --app-config viam-esp32test-main.json --monitor
Downloading micro-RDK release from "https://github.com/viamrobotics/micro-rdk/releases/latest/download/micro-rdk-esp32-server.bin"
✔ Please enter WiFi SSID · [REMOVED]
✔ Please enter WiFi Password · ********
Creating NVS partition with robot id: "[REMOVED]", wifi ssid: "[REMOVED]".
writing entry 1 of 14...
writing entry 2 of 14...
writing entry 3 of 14...
writing entry 4 of 14...
writing entry 5 of 14...
writing entry 6 of 14...
writing entry 7 of 14...
writing entry 8 of 14...
writing entry 9 of 14...
writing entry 10 of 14...
writing entry 11 of 14...
writing entry 12 of 14...
writing entry 13 of 14...
writing entry 14 of 14...
Writing 5 NVS pages
Writing credentials to binary.
Connecting...
Connected. Writing to flash...
Error: EspFlashError(Connection(Timeout(TimedOutCommand { command: Some(FlashDeflateData) })))
```

It seems that's all I get now when I run the script. Is there anything you can suggest @mcvella ? May be a question for Dan M?

**Comment by mcvella** [comment]
I'd post this one in discord and I can have someone jump in to answer.
I don't know the esp32 well but have been told "make sure the cable can support data (and not just charge) and hold down the boot button. It's a pretty standard esp32 flashing thing, not specific to Viam" - not sure if that may be the issue

**Reply by danic85** [reply]
Thanks Matt, it looks like it was an issue with me putting the board into boot mode at the wrong time. For the Arduino IDE it works automatically (on this board at least) or you can enter boot mode when the 'Connecting...' message appears. For this script I needed to press the button a lot sooner, at the point that it is creating the NVS partition.

It looks like i'm connected now so I'll work through more of the guides. Thanks for your help!

**Comment by danic85** [comment]
Hi Esha, I'm following these instructions:

https://www.viam.com/post/build-a-plant-watering-robot-with-webhooks-and-an-esp32

The script exits quickly:

```
2024-01-30 19:27:53,575		INFO	plant-watering-robot (esp32test.py:32)	calling the robot with esp32test-main.n2qdsrpqxj.viam.cloud [removed] [removed]	
2024-01-30 19:27:57,846		INFO	viam.rpc.dial (dial.py:286)	Connecting to socket: /tmp/proxy-sF67z7jJ.sock	
2024-01-30 19:27:58,241		INFO	plant-watering-robot (esp32test.py:91)	closing script after exception ('board', 'board')
```

I don't see anything relevant in the logs for the esp32, and the configuration matches the example (with the comments removed as they broke the JSON)

I'm running a venv with the latest version of the `viam-sdk` installed.

I'm quite happy to debug issues myself, but there isn't a lot to go on here. Any tips to make the process easier?

**Reply by danic85** [reply]
After posting this I unplugged the board and then plugged it back in again, and this time when I ran the script it is working... So I guess that's a 'turn it off and on again' kind of a problem.

The question is still relevant however: how do I get more information on an issue when it occurs, and would we not expect to see something in the app `Logs` log tab?

**Reply by npmenard** [reply]
Hi Dan,

From your description of the issue you ran into I think it could have been caused by Micro-RDK not properly picking up the configuration from app.viam.com. When Micro-RDK boots it reads the config just once so if you make any modification to the config on app.viam.com after that it will not be picked up until the next reboot. 

Micro-RDK only reports errors to app.viam.com when processing the config. After (because of limited resources) we do not send error messages to the app, if you want the get a better "insight" of what if going on you can connect your esp32 to your computer and use `micro-rdk-installer-amd64-linux monitor` to read the logs reported by the esp32 you can also use `--log-file path` parameter to log the console output to a file.

Let me know if you have any questions I'm here to help :)
